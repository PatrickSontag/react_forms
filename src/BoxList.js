import { useState } from 'react';
import { v4 as uuid } from 'uuid';
import './BoxList.css';
import Box from './Box';
import NewBoxForm from './NewBoxForm';

function BoxList() {

    const INITIAL_STATE = [
        { id: uuid(), height: '100px', width: "400px", color: "orange" },
        { id: uuid(), height: '300px', width: "200px", color: "teal" }
      ]
    const [boxes, setBoxes] = useState(INITIAL_STATE);
    const addBox = (newBox) => {
    setBoxes(boxes => [...boxes, { ...newBox, id: uuid() }])
    }

    const removeBox = (id) => {
    setBoxes(boxes.filter(b => b.id !== id))
    }

    const boxInfo = () => {
        console.log("boxes", boxes);
    }
    return (
    <div>
        <h3>Box List</h3>
        <NewBoxForm addBox={addBox} />    
        <div>
        {boxes.map(({ id, height, width, color }) => 
            <Box 
                id={id} boxHeight={height} 
                boxWidth={width} 
                boxColor={color} 
                key={id} 
                removeBox={removeBox}/>)}
        </div>
        <button onClick={boxInfo}>Box List Info</button>
        <div onClick={console.log("div boxes", boxes)}></div>

    </div>
    )

    const [ state, setState ] = useState(null);
    console.log(state);
    return (
        <div className="BoxList">
            <NewBoxForm />
            <Box 
                boxHeight="100px" 
                boxWidth="200px" 
                boxColor="red" 
                />
        </div>
    );
}

export default BoxList;