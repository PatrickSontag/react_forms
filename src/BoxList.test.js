import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

function addTestBox(boxList, height = "2", width = "2", color = "peachpuff") {
    const heightInput = boxList.getByLabelText("Height:");
    const widthInput = boxList.getByLabelText("Width:");
    const backgroundInput = boxList.getByLabelText("Color:");
    fireEvent.change(backgroundInput, { target: { value: color } });
    fireEvent.change(widthInput, { target: { value: width } });
    fireEvent.change(heightInput, { target: { value: height } });
    const button = boxList.getByText("New Box");
    fireEvent.click(button);
  }

it("renders without crashing", function() {
  render(<BoxList />);
});

it("matches snapshot", function() {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("can add a new box", function() {
    const boxList = render(<BoxList />);
  
    // no boxes yet
    expect(boxList.queryByText("X")).not.toBeInTheDocument();

    addTestBox(boxList);
  
    // expect to see a box
    const removeButton = boxList.getByText("X");
    expect(removeButton).toBeInTheDocument();
    expect(removeButton.previousSibling).toHaveStyle(`
        height: 2em;
        width: 2em;
        background-color: peachpuff;
    `);
    // expect form to be empty
    expect(boxList.getAllByDisplayValue("")).toHaveLength(3);

  });

  it("can remove a box", function() {
    const boxList = render(<BoxList />);
    addTestBox(boxList);
  
    const removeButton = boxList.getByText("X");
  
    // click the remove button and the box should be gone
    fireEvent.click(removeButton);
    expect(removeButton).not.toBeInTheDocument();
  });
  