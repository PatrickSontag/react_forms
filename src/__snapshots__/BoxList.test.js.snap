// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Box List
    </h3>
    <form
      id="NewBoxForm"
    >
      <div
        class="form-line"
      >
        <label
          for="height"
        >
          Height:
        </label>
        <input
          id="height"
          name="height"
          value=""
        />
      </div>
      <div
        class="form-line"
      >
        <label
          for="width"
        >
          Width:
        </label>
        <input
          id="width"
          name="width"
          value=""
        />
      </div>
      <div
        class="form-line"
      >
        <label
          for="color"
        >
          Color:
        </label>
        <input
          id="color"
          name="color"
          value=""
        />
      </div>
      <button
        class="form-button"
      >
        New Box
      </button>
    </form>
    <div>
      <div
        class="Box"
      >
        <div
          class="box"
          style="height: 10em; width: 10em; background-color: orange;"
        />
        <button>
          X
        </button>
      </div>
      <div
        class="Box"
      >
        <div
          class="box"
          style="height: 5em; width: 20em; background-color: teal;"
        />
        <button>
          X
        </button>
      </div>
    </div>
    <div />
  </div>
</DocumentFragment>
`;
